### PR TITLE
Deprecate GDP Reclassify Transformation in Favor of Using References

### PR DESCRIPTION
**Pull request by emma58**
## Summary/Motivation:

Previously, `TransformationFactory('gdp.bigm').apply_to(m.block)` and `TransformationFactory('gdp.bigm').apply_to(m, targets=m.block)` behaved differently because of the hack to expose the `indicator_var`s and other local variables to the writers. This is because the first automatically called the `gdp.reclassify` transformation whereas the second did not, since we have no way of knowing if someone expects their model to be completely transformed after using targets.

This PR deprecates `gdp.reclassify` and instead has the bigm and hull transformations create References to locally declared variables, which are stored on their own sub-block of the Disjunct's transformation Block. This exposes the local variables to the writers without ever changing the ctype of the `Disjunct`.

## Changes proposed in this PR:
- Modifies `gdp.bigm` and `gdp.hull` to add References to all locally declared Vars on the Disjunct transformation block.
- Adds `assert_model_algebraic` function to gdp.util to take the place of the validation we were doing in `gdp.reclassify` to check that the model was indeed completely transformed.
- Moves `hull` disaggregated variables to their own sub-block of the Disjunct transformation block. (This is peripheral to this PR, but means we have complete control over names on the Disjunct transformation block again, which is nice.)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/gdp/plugins/gdp_var_mover.py [hunk 1]
```diff
@@ -22,39 +22,13 @@
 from pyomo.core import TraversalStrategy, TransformationFactory
 from pyomo.core.base.indexed_component import ActiveIndexedComponent
 from pyomo.common.deprecation import deprecated
+from pyomo.common.modeling import unique_component_name
+from pyomo.common.config import ConfigBlock, ConfigValue
```
Comment (jsiirola) [inline pyomo/gdp/plugins/gdp_var_mover.py:26]: This import is not used

File: pyomo/gdp/plugins/gdp_var_mover.py [hunk 2]
```diff
@@ -64,7 +38,13 @@ class HACK_GDP_Disjunct_Reclassifier(Transformation):
...
 
     """
-
+    @deprecated(msg="The gdp.reclasify transformation has been deprecated in "
+                "favor of the gdp transformations creating References to "
+                "variables local to each Disjunct during the transformation. "
+                "Validation that the model has been completely transformed "
+                "to an algebraic model has been moved to the "
+                "assert_model_algebraic function in gdp.util.",
+                version='5.6.10')
```
Comment (jsiirola) [inline pyomo/gdp/plugins/gdp_var_mover.py:47]: The next release will be 5.7 (not 5.6.10)

**Comment by blnicho** [comment]
@jsiirola is this ready to merge?

**Comment by jsiirola** [comment]
Yes.

**Pull request merged**